suprb/optimizer/solution/ga/base.py

```diff
@@ -62,13 +64,43 @@ def __init__(
         self.crossover = crossover
         self.selection = selection
         self.elitist_ratio = elitist_ratio
+        self.mutation_rate = mutation_rate
+        self.crossover_rate = crossover_rate
+
+    def adjust_rates(self):
+        pass
+
+    def fitness_calculation(self):
+        pass
+
+    def crossover_children(self, parent_pairs):
+        return list(
+            flatten(
+                [
+                    (
+                        self.crossover(A, B, self.crossover_rate, random_state=self.random_state_),
+                        self.crossover(B, A, self.crossover_rate, random_state=self.random_state_),
+                    )
+                    for A, B in parent_pairs
+                ]
+            )
+        )
+
+    def mutation_children(self, X, y, children):
+        return [self.mutation(child, self.mutation_rate, random_state=self.random_state_) for child in children]
 
     def _optimize(self, X: np.ndarray, y: np.ndarray):
+        assert self.population_size % 2 == 0
         self.fit_population(X, y)
 
         self.n_elitists_ = int(self.population_size * self.elitist_ratio)
 
         for _ in range(self.n_iter):
+            self.fitness_calculation()
+
+            # Adjust Rates for current fitness
```

Comment:
unclear to me what this does from the name. should be renamed to reflect that it invokes the self-adaptive component

Reply:
It's the fitness calculation for mutation and crossover for `SelfAdaptingGeneticAlgorithm2`, will move this into the saga directory as well

Reply:
I feel like it has been added in the other file and can just go here?

Reply:
Yes removed